src/rust/engine/src/context.rs

```diff
@@ -349,6 +355,23 @@ impl Core {
       || exec_strategy_opts.remote_cache_read
       || exec_strategy_opts.remote_cache_write;
 
+    // If the remote store and remote execution server are the same (address and headers),
+    // then share the capabilities cache between them to avoid duplicate GetCapabilities calls.
+    let capabilities_cell_opt = match (
+      need_remote_store,
+      remoting_opts.execution_address.as_ref(),
+      remoting_opts.store_address.as_ref(),
+      &remoting_opts.execution_headers,
+      &remoting_opts.store_headers,
+    ) {
+      (true, Some(execution_address), Some(store_address), execution_headers, store_headers)
+        if execution_address == store_address && execution_headers == store_headers =>
+      {
+        Some(Arc::new(DoubleCheckedCell::new()))
+      }
+      _ => None,
+    };
```

Comment:
I think less ceremony to directly use `if else`? `if need_remote_store && remoting_opts.execution_address ==   remoting_opts.store_address && ...`

Reply:
Ya. Fixed.